_posts/2016-09-00-edsls-part-1.md

```diff
@@ -0,0 +1,76 @@
+---
+layout: post
+title: It's programs all the way down
+
+meta:
+  nav: blog
+  author: adelbertc
+  pygments: true
+---
+
+*This is the first of a series of articles on "EDSLs in Scala."*
+
+Embedded domain specific languages (EDSLs) are powerful tools for
+abstracting complexities such as effects and business logic from our
+programs. Instead of mixing ad-hoc error handling, database access, and web
+calls through our code, we isolate each domain into a little language. These
+little languages can then be used to write "mini-programs" describing for
+example, how to create a web page for a user.
+
+Our program then becomes a composition of mini-programs, and running our
+program becomes interpreting these mini-programs into actions. This is
+analogous to running an interpreter, itself a program, which turns code
+into actions.
+
+The following illustrates what an EDSL might look like in Scala.
+
+```scala
+// An embedded program for fetching data for a user
+def process(id: UserId): Program[Page] = for {
```

Comment:
I should probably add clarification that this does not actually do anything, it's only built up a structure.